File: base_java/src/main/java/com/hesheng1024/base_java/adapter/BaseRecyclerAdapter.java
package com.hesheng1024.base_java.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

/**
 * @author hesheng1024
 * @date 2019/5/29 17:49
 */
public abstract class BaseRecyclerAdapter<T> extends RecyclerView.Adapter<BaseViewHolder> {

    private final List<T> mDataList = new ArrayList<>();
    private BaseItemClickListener mListener = null;

    public BaseRecyclerAdapter() {
    }

    public BaseRecyclerAdapter(@NonNull List<T> datas) {
        addList(datas);
    }

    public void add(@NonNull T data) {
        mDataList.add(data);
        notifyDataSetChanged();
    }

    public void add(int position, @NonNull T data) {
        mDataList.add(position, data);
        notifyDataSetChanged();
    }

    public void addList(@NonNull List<T> dataList) {
        mDataList.addAll(dataList);
        notifyDataSetChanged();
    }

    public void remove(int position) {
        mDataList.remove(position);
        notifyDataSetChanged();
    }

    public void remove(@NonNull T data) {
        mDataList.remove(data);
        notifyDataSetChanged();
    }

    public void clear() {
        mDataList.clear();
        notifyDataSetChanged();
    }

    @Nullable
    public T get(int position) {
        return mDataList.get(position);
    }


    public void setClickListener(BaseItemClickListener listener) {
        this.mListener = listener;
    }

    @Nullable
    protected BaseItemClickListener getListener() {
        return mListener;
    }

    @NonNull
    @Override
    public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new BaseViewHolder(LayoutInflater
                .from(parent.getContext())
                .inflate(getItemLayoutId(), parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder holder, int position) {
        bindHolder(holder, position);
    }

    @Override
    public int getItemCount() {
        return mDataList.isEmpty() ? 0 : mDataList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return super.getItemViewType(position);
    }

    /**
     * item layout id
     * @return resId
     */
    @LayoutRes
    protected abstract int getItemLayoutId();

    /**
     * 绑定ViewHolder
     *
     * @param holder: view holder
     * @param position: position in adapter
     */
    protected abstract void bindHolder(BaseViewHolder holder, int position);
}

File: base_kotlin/src/main/java/com/hesheng1024/base_kotlin/adapter/BaseRecyclerAdapter.kt
package com.hesheng1024.base.base.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.recyclerview.widget.RecyclerView

/**
 *
 * @author hesheng1024
 * @date 2020/2/8 11:28
 */
abstract class BaseRecyclerAdapter<T>(dataList: List<T>? = null) : RecyclerView.Adapter<BaseViewHolder>() {

    private val mDataList: MutableList<T> = ArrayList()
    private var mListener: IBaseIClickListener? = null

    init {
        if (!dataList.isNullOrEmpty()) {
            this.addList(dataList)
        }
    }

    open fun setList(@NonNull data: List<T>) {
        mDataList.clear()
        addList(data)
    }

    open fun add(@NonNull data: T) {
        mDataList.add(data)
        notifyDataSetChanged()
    }

    open fun add(position: Int, @NonNull data: T) {
        mDataList.add(position, data)
        notifyDataSetChanged()
    }

    open fun addList(@NonNull data: List<T>) {
        mDataList.addAll(data)
        notifyDataSetChanged()
    }

    open fun remove(position: Int) {
        mDataList.removeAt(position)
        notifyDataSetChanged()
    }

    open fun remove(@NonNull data: T) {
        mDataList.remove(data)
        notifyDataSetChanged()
    }

    open fun clear() {
        mDataList.clear()
        notifyDataSetChanged()
    }

    @Nullable
    open fun get(position: Int): T {
        return mDataList[position]
    }

    open fun getList(): List<T> {
        return mDataList
    }

    open fun setClickListener(listener: IBaseIClickListener) {
        this.mListener = listener
    }

    protected fun getListener() = mListener

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        return BaseViewHolder(LayoutInflater.from(parent.context).inflate(getItemLayoutId(), parent, false))
    }

    override fun getItemCount(): Int {
        return if (mDataList.isNullOrEmpty()) 0 else mDataList.size
    }

    /**
     * item layout_block_edit_text id
     * @return resId
     */
    @LayoutRes
    protected abstract fun getItemLayoutId(): Int
}